drivers/platform/aducm3029/spi.c

```diff
@@ -0,0 +1,232 @@
+	if (NULL == desc)
+		return NULL;
+
+	extra = malloc(sizeof(struct adicup_spi_desc));
+	if (NULL == extra) {
+		free(desc);
+		return NULL;
+	}
+	desc->extra = extra;
+
+	extra->buffer = malloc(ADI_SPI_MEMORY_SIZE);
+	if (NULL == extra->buffer) {
+		free(extra);
+		free(desc);
+		return NULL;
+	}
+
+	return desc;
+}
+
+/**
+ * @brief Deallocates the memory needed for the descriptor
+ * @param desc - Descriptor to be deallocated
+ */
+void free_desc_mem(struct spi_desc *desc)
+{
+	free(((struct adicup_spi_desc*)(desc->extra))->buffer);
+	free(desc->extra);
+	free(desc);
+}
+
+/**
+ * @brief Initialize the SPI communication peripheral.
+ * @param desc - The SPI descriptor.
+ * @param init_param - The structure that contains the SPI parameters.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t spi_init(struct spi_desc **desc,
+		 const struct spi_init_param *param)
+{
+	ADI_SPI_RESULT			spi_ret;
+	struct adicup_spi_desc		*adicup_desc;
+	struct adicup_spi_init_param	*adicup_init_param;
+
+
+	if (desc == NULL || param == NULL)
+		goto fail_no_mem;
+
+	/* Verify if the SPI is already activated */
+	adicup_init_param = param->extra;
+	if (active_spi & (1 << adicup_init_param->spi_channel))
+		goto fail_mem;
+
+	*desc = alloc_data_mem(param);
+	if (*desc == NULL)
+		goto fail_no_mem;
+	(*desc)->chip_select = param->chip_select;
+	(*desc)->max_speed_hz = param->max_speed_hz;
+	(*desc)->mode = param->mode;
+	adicup_desc = (*desc)->extra;
+	adicup_desc->spi_channel = adicup_init_param->spi_channel;
+	adicup_desc->master_mode = adicup_init_param->master_mode;
+	adicup_desc->dma = false;
+	adicup_desc->half_duplex = false;
+
+	/* Configure the SPI with user parameters*/
+	spi_ret = adi_spi_Open(adicup_init_param->spi_channel,adicup_desc->buffer,
+			       ADI_SPI_MEMORY_SIZE, &adicup_desc->spi_handle);
+	if (spi_ret != ADI_SPI_SUCCESS)
+		goto fail_mem;
+	spi_ret = adi_spi_SetBitrate(adicup_desc->spi_handle, param->max_speed_hz);
+	if (spi_ret != ADI_SPI_SUCCESS)
+		goto fail_mem;
+	spi_ret = adi_spi_SetChipSelect(adicup_desc->spi_handle, param->chip_select);
+	if (spi_ret != ADI_SPI_SUCCESS)
+		goto fail_mem;
+	spi_ret = adi_spi_SetClockPhase(adicup_desc->spi_handle,
+					param->mode & SPI_CPHA);
+	if (spi_ret != ADI_SPI_SUCCESS)
+		goto fail_mem;
+	spi_ret = adi_spi_SetClockPolarity(adicup_desc->spi_handle,
+					   param->mode & SPI_CPOL);
+	if (spi_ret != ADI_SPI_SUCCESS)
+		goto fail_mem;
+	spi_ret = adi_spi_SetMasterMode(adicup_desc->spi_handle,
+					adicup_init_param->master_mode);
+	if (spi_ret != ADI_SPI_SUCCESS)
+		goto fail_mem;
+	spi_ret = adi_spi_SetContinuousMode(adicup_desc->spi_handle,
+					    adicup_init_param->continuous_mode);
+	if (spi_ret != ADI_SPI_SUCCESS)
+		goto fail_mem;
+
+	/* SPI configured. Set it as activated */
+	active_spi |= (1 << adicup_init_param->spi_channel);
+
+	spi_ret = SUCCESS;
+	goto succes;
+fail_mem:
+	free_desc_mem(*desc);
+fail_no_mem:
+	spi_ret = FAILURE;
+succes:
```

Comment:
The label is a bit confusing since you go there even if the return value is FAILURE